File: demos/pax-britannica/pax-britannica-android/assets/data/scripts/ship.lua
local v2 = require 'dokidoki.v2'

assert(player, 'missing player argument')
assert(turn_speed, 'missing turn_speed argument')
assert(accel, 'missing accel argument')
assert(hit_points, 'missing hit_points argument')

local max_hit_points = hit_points
self.sprite.color = {1.0, 1.0, 1.0}

if sprites_table then
  self.sprite.image = assert(game.resources[sprites_table][player])
end

velocity = v2(0, 0)

-- 1 for left, -1 for right
function turn(direction, amount)
  amount = amount or 1
  self.transform.facing =
    v2.norm(v2.rotate(self.transform.facing, turn_speed * direction * amount))
end

function thrust(amount)
  amount = amount or 1
  velocity = velocity + self.transform.facing * accel * amount
end

local function random_point_on_ship()
  -- assumes the ship is a rectangle
  local a = self.collision.poly.vertices[1]
  local ab = self.collision.poly.vertices[2] - a
  local ac = self.collision.poly.vertices[4] - a
  return self.transform.pos +
         v2.rotate_to(self.transform.facing,
                      a + ab * math.random() + ac * math.random())
end

function update()
  velocity = velocity * 0.97
  self.transform.pos = self.transform.pos + velocity
  if hit_points <= 0 then destruct() end

  if math.random() < v2.mag(velocity) / 10 then
    game.particles.add_bubble(random_point_on_ship())
  end

  -- debug
  game.tracing.trace_bar(self.transform.pos + v2(0, 10),
                         hit_points / max_hit_points)
end

local function go_towards_or_away(target_pos, force_thrust, is_away)
  local target_direction = target_pos - self.transform.pos
  if is_away then target_direction = -target_direction end

  if v2.cross(self.transform.facing, target_direction) > 0 then
    turn(1)
  else
    turn(-1)
  end
  if force_thrust or v2.dot(self.transform.facing, target_direction) > 0 then
    thrust()
  end
end

function health_percentage()
  return math.max(hit_points / max_hit_points, 0)
end

function damage(amount)
  hit_points = math.max(hit_points - amount, 0)
end

function destruct()
  if self.factory_ai then
    factory_destruct()
  else
    game.particles.explode(self)
    game.log.record_death(self.blueprint)
    self.dead = true
  end
end

local death_counter = 100
local next_explosion = 10
local opacity = 0.6

function factory_destruct()
  if death_counter > 0 then
    self.production.halt_production = true
    self.sprite.color = {self.sprite.color[1], self.sprite.color[2], self.sprite.color[3], math.max(0, opacity)}
    opacity = opacity - 0.006
    if death_counter % next_explosion == 0 then
      game.particles.explode(self, random_point_on_ship())
      next_explosion = math.random(6,15)
    end
    death_counter = death_counter - 1
  else
    for i = 1,5 do
      random = (v2.random() + v2.random()) * 20
      game.particles.explode(self, self.transform.pos + random)
    end
    game.log.record_death(self.blueprint)
    self.dead = true
  end
end

-- automatically thrusts and turns according to the target
function go_towards(target_pos, force_thrust)
  go_towards_or_away(target_pos, force_thrust, false)
end

function go_away(target_pos, force_thrust)
  go_towards_or_away(target_pos, force_thrust, true)
end

File: demos/pax-britannica/pax-britannica/src/de/swagner/paxbritannica/Ship.java
package de.swagner.paxbritannica;

import java.util.ArrayList;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;

import de.swagner.paxbritannica.factory.FactoryProduction;

public class Ship extends Sprite {

	protected float amount = 1.0f;

	protected float turnSpeed = 1.0f;
	protected float accel = 0.0f;
	protected float hitPoints = 0;

	protected float maxHitPoints = 0;

	private float delta = 0.0f;

	public float aliveTime = 0.0f;

	public Vector2 position = new Vector2();
	public Vector2 velocity = new Vector2();
	public Vector2 facing = new Vector2();
	
	public Vector2 collisionCenter = new Vector2();
	public ArrayList<Vector2> collisionPoints = new ArrayList<Vector2>();

	public boolean alive = true;

	private float deathCounter = 50f;
	private float nextExplosion = 10f;
	private float opacity = 5.0f;

	public int id = 0;

	public Ship(int id, Vector2 position, Vector2 facing) {
		super();

		this.id = id;

		this.position.set(position);
		this.facing.set(facing);
		
		collisionPoints.clear();
		collisionPoints.add(new Vector2());
		collisionPoints.add(new Vector2());
		collisionPoints.add(new Vector2());
		collisionPoints.add(new Vector2());

		this.setOrigin(this.getWidth() / 2.f, this.getHeight() / 2.f);
	}

	@Override
	public void draw(SpriteBatch spriteBatch) {
		delta = Math.min(0.06f, Gdx.graphics.getDeltaTime());
		
		aliveTime += delta;
		collisionPoints.get(0).set( this.getVertices()[0], this.getVertices()[1]);
		collisionPoints.get(1).set( this.getVertices()[5], this.getVertices()[6]);
		collisionPoints.get(2).set( this.getVertices()[10], this.getVertices()[11]);
		collisionPoints.get(3).set( this.getVertices()[15], this.getVertices()[16]);
		
		collisionCenter.set(collisionPoints.get(2)).scl(0.5f).add(collisionPoints.get(0));

		velocity.scl( (float) Math.pow(0.97f, delta * 30.f));
		position.add(velocity.x * delta, velocity.y * delta);
		
		this.setRotation(facing.angle());
		this.setPosition(position.x, position.y);

		if (!(this instanceof Bullet) && hitPoints <= 0)
			destruct();
		if (MathUtils.random() < velocity.len() / 900.f) {
			GameInstance.getInstance().bubbleParticles.addParticle(randomPointOnShip());
		}
		super.draw(spriteBatch);
	}

	public void turn(float direction) {
		delta = Math.min(0.06f, Gdx.graphics.getDeltaTime());
		
		facing.rotate(direction * turnSpeed * delta).nor();
	}

	public void thrust() {
		delta = Math.min(0.06f, Gdx.graphics.getDeltaTime());
		
		velocity.add(facing.x * accel * delta, facing.y * accel * delta);
	}
	
	public void thrust(float amount) {
		delta = Math.min(0.06f, Gdx.graphics.getDeltaTime());
		
		velocity.add(facing.x * accel * delta, facing.y * accel * amount * delta);
	}

	public Vector2 randomPointOnShip() {
		return new Vector2(collisionCenter.x + MathUtils.random(-this.getWidth() / 2, this.getWidth() / 2), collisionCenter.y
				+ MathUtils.random(-this.getHeight() / 2, this.getHeight() / 2));
	}

	/*
	 * Scratch space for computing a target's direction. This is safe (as a static) because
	 * its only used in goTowardsOrAway which is only called on the render thread (via update).
	 */
	private static final Vector2 target_direction = new Vector2(); 

	public void goTowardsOrAway(Vector2 targetPos, boolean forceThrust, boolean isAway) {
		target_direction.set(targetPos).sub(collisionCenter);
		if (isAway) {
			target_direction.scl(-1);
		}

		if (facing.crs(target_direction) > 0) {
			turn(1);
		} else {
			turn(-1);
		}

		if (forceThrust || facing.dot(target_direction) > 0) {
			thrust();
		}
	}

	public float healthPercentage() {
		return Math.max(hitPoints / maxHitPoints, 0);
	}

	public void damage(float amount) {
		hitPoints = Math.max(hitPoints - amount, 0);
	}

	public void destruct() {
		if (this instanceof FactoryProduction) {
			factoryDestruct();
		} else {
			GameInstance.getInstance().explode(this);
			alive = false;
			for (Ship factory : GameInstance.getInstance().factorys) {
				if(factory instanceof FactoryProduction && factory.id == this.id) ((FactoryProduction) factory).ownShips--;
			}
		}
	}

	public void factoryDestruct() {
		delta = Math.min(0.06f, Gdx.graphics.getDeltaTime());

		if (deathCounter > 0) {
			((FactoryProduction) this).production.halt_production = true;
			this.setColor(1, 1, 1, Math.min(1, opacity));
			opacity -= 1 * delta;
			if (Math.floor(deathCounter) % nextExplosion == 0) {
				GameInstance.getInstance().explode(this, randomPointOnShip());
				nextExplosion = MathUtils.random(2, 6);
			}
			deathCounter -= 10 * delta;
		} else {
			for (int i = 1; i <= 10; ++i) {
				GameInstance.getInstance().explode(this, randomPointOnShip());
			}
			alive = false;
		}
	}

	// automatically thrusts and turns according to the target
	public void goTowards(Vector2 targetPos, boolean forceThrust) {
		goTowardsOrAway(targetPos, forceThrust, false);
	}

	public void goAway(Vector2 targetPos, boolean forceThrust) {
		goTowardsOrAway(targetPos, forceThrust, true);
	}

}
